File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
while 1:
    s = input()
    stack = []
    result = True

    if s == '.':
        break
    
    for i in s:
        if i == '(' or i == '[':
            stack.append(i)
        elif i == ')':
            if stack and stack[-1] == '(':
                stack.pop()
            else:
                result = False
                break
        elif i == ']':
            if stack and stack[-1] == '[':
                stack.pop()
            else:
                result = False
                break
    
    if stack or not result:
        print('no')
    else:
        print('yes')

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.java
import java.util.Stack;
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        while (true) {
            boolean flag = true;
            String s = sc.nextLine();
            Stack<Character> stack = new Stack<>();
            
            if (s.equals(".")) break;
            
            for (char c : s.toCharArray()) {
                if (c == '(' || c == '[') stack.push(c);
                else if (c == ')') {
                    if (!stack.isEmpty() && stack.peek() == '(') stack.pop();
                    else {
                        flag = false;
                        break;
                    }
                } 
                else if (c == ']') {
                    if (!stack.isEmpty() && stack.peek() == '[') stack.pop();
                    else {
                        flag = false;
                        break;
                    }
                }
            }
            
            if (!stack.isEmpty()) flag = false;
            System.out.println(flag ? "yes" : "no");
        }
    }
}